service/frontend/templates/ratelimited.tmpl

```diff
@@ -0,0 +1,117 @@
+{{$ratelimitTypeMap = set $ratelimitTypeMap "ListTaskListPartitions" "ratelimitTypeUser"}}
+{{$ratelimitTypeMap = set $ratelimitTypeMap "GetTaskListsByDomain" "ratelimitTypeUser"}}
+{{$ratelimitTypeMap = set $ratelimitTypeMap "RefreshWorkflowTasks" "ratelimitTypeUser"}}
+
+{{$domainIDAPIs := list "RecordActivityTaskHeartbeat" "RespondActivityTaskCanceled" "RespondActivityTaskCompleted" "RespondActivityTaskFailed" "RespondDecisionTaskCompleted" "RespondDecisionTaskFailed" "RespondQueryTaskCompleted"}}
+{{$queryTaskTokenAPIs := list "RespondQueryTaskCompleted"}}
+{{$nonBlockingAPIs := list "RecordActivityTaskHeartbeat" "RecordActivityTaskHeartbeatByID" "RespondActivityTaskCompleted" "RespondActivityTaskCompletedByID" "RespondActivityTaskFailed" "RespondActivityTaskFailedByID" "RespondActivityTaskCanceled" "RespondActivityTaskCanceledByID" "RespondDecisionTaskCompleted" "RespondDecisionTaskFailed" "RespondQueryTaskCompleted" "ResetStickyTaskList"}}
+
+{{$interfaceName := .Interface.Name}}
+{{$handlerName := (index .Vars "handler")}}
+{{ $decorator := (printf "%s%s" (down $handlerName) $interfaceName) }}
+{{ $Decorator := (printf "%s%s" $handlerName $interfaceName) }}
+
+// {{$decorator}} implements {{.Interface.Type}} interface instrumented with rate limiter.
+type {{$decorator}} struct {
+    wrapped     {{.Interface.Type}}
+    tokenSerializer common.TaskTokenSerializer
+    domainCache cache.DomainCache
+    userRateLimiter quotas.Policy
+    workerRateLimiter quotas.Policy
+    visibilityRateLimiter quotas.Policy
+}
+
+// New{{$Decorator}} creates a new instance of {{$interfaceName}} with ratelimiter.
+func New{{$Decorator}}(
+    wrapped {{.Interface.Type}},
+    domainCache cache.DomainCache,
+    userRateLimiter quotas.Policy,
+    workerRateLimiter quotas.Policy,
+    visibilityRateLimiter quotas.Policy,
+) {{.Interface.Type}} {
+    return &{{$decorator}}{
+        wrapped: wrapped,
+        tokenSerializer: common.NewJSONTaskTokenSerializer(),
+        domainCache: domainCache,
+        userRateLimiter: userRateLimiter,
+        workerRateLimiter: workerRateLimiter,
+        visibilityRateLimiter: visibilityRateLimiter,
+    }
+}
+
+{{range $method := .Interface.Methods}}
+func (h *{{$decorator}}) {{$method.Declaration}} {
+    {{- if hasKey $ratelimitTypeMap $method.Name}}
+    {{- $ratelimitType := get $ratelimitTypeMap $method.Name}}
+    {{- $domain := printf "%s.GetDomain()" (index $method.Params 1).Name}}
+    {{- $rateLimitHelper := "allowDomain"}}
+    {{- if has $method.Name $domainIDAPIs}}
+    {{- $domain = "domainName"}}
+    {{- $rateLimitHelper := "allowDomainID"}}
+    {{- if has $method.Name $queryTaskTokenAPIs}}
+    token, err := h.tokenSerializer.DeserializeQueryTaskToken({{(index $method.Params 1).Name}}.TaskToken)
+    {{- else}}
+    token, err := h.tokenSerializer.Deserialize({{(index $method.Params 1).Name}}.TaskToken)
+    {{- end}}
```

Comment:
fwiw: you can indent the code lines on all of these as deeply as you want, the formatting step will fix the indentation.

it won't remove empty lines, so the `{{- ` is still useful/desirable, but indents help a fair bit with where the if/ends are paired.